crates/sui-core/src/execution_cache/memory_cache.rs

```diff
@@ -0,0 +1,1143 @@
+    ) -> BoxFuture<'a, SuiResult> {
+        todo!()
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    fn write_transaction_outputs(
+        &self,
+        epoch_id: EpochId,
+        tx_outputs: Arc<TransactionOutputs>,
+    ) -> BoxFuture<'_, SuiResult> {
+        let TransactionOutputs {
+            transaction,
+            effects,
+            markers,
+            written,
+            deleted,
+            wrapped,
+            events,
+            ..
+        } = &*tx_outputs;
+
+        // Update all markers
+        for (object_key, marker_value) in markers.iter() {
+            self.dirty
+                .markers
+                .entry((epoch_id, object_key.0))
+                .or_default()
+                .value_mut()
+                .insert(object_key.1, *marker_value);
+        }
+
+        // Write children before parents to ensure that readers do not observe a parent object
+        // before its most recent children are visible.
+        for (object_id, object) in written.iter() {
+            if object.is_child_object() {
+                self.write_object(object_id, object);
+            }
+        }
+        for (object_id, object) in written.iter() {
+            if !object.is_child_object() {
+                self.write_object(object_id, object);
+                if object.is_package() {
+                    self.cached
+                        .packages
+                        .insert(*object_id, PackageObject::new(object.clone()));
+                }
+            }
+        }
+
+        for ObjectKey(id, version) in deleted.iter() {
+            self.write_deleted_tombstone(id, *version);
+        }
+        for ObjectKey(id, version) in wrapped.iter() {
+            self.write_wrapped_tombstone(id, *version);
+        }
+
+        let tx_digest = *transaction.digest();
+        let effects_digest = effects.digest();
+
+        self.dirty
+            .transaction_effects
+            .insert(effects_digest, effects.clone());
+
+        match self.dirty.transaction_events.entry(events.digest()) {
+            DashMapEntry::Occupied(mut occupied) => {
+                occupied.get_mut().0.insert(tx_digest);
+            }
+            DashMapEntry::Vacant(entry) => {
+                let mut txns = BTreeSet::new();
+                txns.insert(tx_digest);
+                entry.insert((txns, events.clone()));
+            }
+        }
+
+        self.dirty
+            .executed_effects_digests
+            .insert(tx_digest, effects_digest);
+
+        self.dirty
+            .pending_transaction_writes
+            .insert(tx_digest, tx_outputs);
+
+        self.executed_effects_digests_notify_read
+            .notify(&tx_digest, &effects_digest);
+
+        if let Some(metrics) = &self.metrics {
+            metrics
+                .pending_notify_read
+                .set(self.executed_effects_digests_notify_read.num_pending() as i64);
+        }
+
+        std::future::ready(Ok(())).boxed()
```

Comment:
Should this method also call `commit_transaction_outputs` right now (but eventually move to checkpoint finalization)? I might have missed where the data is committed to DB.

Reply:
So far the data is not committed to the db at all. (None of this code will run yet, even after it is merged. We are still using PassthroughCache). The commits will be driven by CheckpointExecutor in a forthcoming PR